### DISCUSSION TITLE
Per user default chat colour

### DISCUSSION THREAD

**Discussion by e101stella** [Zrips/CMI#3516] [Feature Requests]
**Description of feature/improvement request:**
Basically as title, allow for the setup of per user chat colour (their text in chat) with the possibility of also allowing players to change their own chat colour.

There are other plugins, such as ChatColor2, which simply add a &colour at the start of any chat message, which works in most cases, however these plugins have the problem of not working as intended with text editors such as /cmi aliaseditor, /cmi editctext, or /npc text in citizens. 

---
### Detailed explanation of what it should do, when and in what situation: ###
This should be an extension of CMI chat formatting, that prepends a chat colour code to a user's message in public chat formatting _(possibility of expanding it to messages, shouts, etc, only particularly useful for public chat in my case)_, but only does so in the instance of a public message and doesn't prepend it for any text message. 

In terms of storing a users chat colour, this would likely be an additional field in the user database (or an extra table), with a default colour specified in the config for users who haven't had their chat colour changed.

In the case of allowing players to change their own text colour, a command such as /cmi chatcolour could bring up a menu of available chat colours and clicking on the colour would change a users chat colour, or alternatively a player could just specify the colour or colour code doing /cmi chatcolour [colour]. It would also be highly beneficial to be able to implement a cooldown between colour changes which I understand could be done with the use of the built-in cooldowns module, but could possibly benefit from a proper usage if the former method of changing colour is used.

### Basic explanation as to why this feature is needed and/or useful to have: ###
This feature would enable the ability to distinguish staff/admin messages from normal player messages easily, as well as adding a EULA compliant donation perk for players.
It would also be (if implemented as described) a robust tool that would help futher integrate CMI's chat manager into other components.

**Comment by e101stella** [comment]
It would be possible to manually setup a bunch of different groups each with a different message colour formatting, and then manage that through permissions, however this is a far clunkier solution to the problem, and limit the ability to modify how the user is presented for different things.

**Comment by CloudeLecaw** [comment]
You can define a by user default color if you just create you chatformat to it. Simply usw the suffix or prefix after the Meassage and give the users the speziffic colorcode as suffix or prefix.  If this is counted for a group you can defibe the group defaults by setting up not the regular chat format, use the grouped one and give each group the right permission for its chatformat. There are no limits like you descripte it. So an exzra Option like that is nit really needed. Use the prefix and suffix + the grouped Format, thats all you need. Also renember your permission System can definie this as global and also per world. Also for each player if that really is needed by u. So i see no reason why this should be added.
 Allowing users to choose their own chat color? Whats next Emoticons?.. Well ok the useage of Emoticons in chats with a placeholder or something that can be used in chat like in whatsapp and co in a menue the use would be cool ;D, but maybe its not possible to design this by Minecraft way how the chatline itself cant be designed

**Comment by e101stella** [comment]
If I understand correctly, you're suggesting using the grouped formats for this purpose? Unless I'm missing the use of some other placeholder, I'm already using both prefix and suffix for different ranks, prefix for donator or staff ranks, and suffix for ranking up.
 
That is a possibility, as I commented on after the main suggestion, however it doesn't have the same flexibility as the entirety of the suggestion that I made. It would also complicate the process of being able to change a players colour, as it would require manual permission changes as opposed to built in command access.

As for why this should be incorporated into CMI, I did specifically outline why in my suggestion, it's a feature offered by some other plugins, but breaks text editors because of how they implement it. 

Just as a side note, emoji's are offered by at least one plugin I believe, but a slippery slope argument is typically not a good example of why something shouldn't be done, and as far as I know CMI's chat formatting plays nicely with emjoi's although I haven't actually tested it myself.

Edit: I also specifically outlined it as a possible donator perk that would have restrictions based on access to public chat colour permissions, as well as putting a cooldown on it so it's not something that can be spammed. Also if you were to give permission for cmi.color.publicmessage.x or whatever it is you'd allow people to use normal colour codes in their chat, and I've already given donators the ability to have a coloured nickname, is that something that shouldn't be allowed at all as well? 

_P.S On the topic of slippery slope arguments, they're typically used to say "Well why should we let them do this, they'll ask for something more extreme which we definitely don't want", which is not an argument against doing that thing, so much as it is against the idea of giving more options and wanting to maintain restrictions. (Most commonly used in politics, generally as a reason why human rights shouldn't apply for every human on some trivial manner such as race or sexual orientation)_

**Comment by CloudeLecaw** [comment]
Hm you right about the flexibility if your using suffix and prefix allready. For the things you wrote, im wondering a bit why your plugin for this does not has this placeholders,... But ok.. Back to theme.. there are just 1,2 server in this world that do such a thing that you suggest. Think of the colotcais you would have if you have bigger Servers where 20 Player are all chatting at once in different colors... As more colorrights you give as sooner your eyes start bleeding from thus chat-Clown-Festival. Most definie color by groups so most player use the same color. This means one color stay mostly in chat. If you need it, you can setips donators perk, give some timed permission to groups/player so use some colorcodes. In cmi Colorcode permissions can be set for all kind of meassages  (chat, public, me, shout, private, tilte, others.. You find the possible permissions inside the chat chatconfigs and on the cmi wiki) Players can use /cmi colors (/,colors if actived) if you gave permission too, to see all possible colorcodes. If they can not use i lt because of missing permission, its up to your config if the colorcodes get prinet of chat or cuttet out of the meassages. 

Maybe it would be also a nice Feature if a player just would see the lines there has permission for :)
